### PR TITLE
Convert url/Slug to TS

### PR DESCRIPTION
**Pull request by keichinger**

### CODE REVIEW

File: url/Slug.ts [hunk 1]
```diff
@@ -1,27 +1,14 @@
...
-     * Constructs a new instance
-     *
-     * @param {Array.<Array.<*>>} transforms
-     * @param {?RegExp} sanitize
+     * The sanitizing regexp, that removes all unwanted characters
      */
-    constructor (transforms = [], sanitize = null)
+    private readonly sanitize : RegExp;
+
+    constructor (transforms : Array<[RegExp, string]> = [], sanitize : null|RegExp= null)
```
Comment (apfelbox) [inline url/Slug.ts]: This type is actually wrong, and explains why you were able to remove the `flags` check further below.

The correct type is: `Array<[RegExp|string, string]>`
Also: please add a named type for that in this file.

File: url/Slug.ts [hunk 2]
```diff
@@ -1,27 +1,14 @@
-
-
-export default class Slug
+export class Slug
```
[the remark this answers is not in the document]
Comment (keichinger) [reply url/Slug.ts]: Oh yes, you're absolute right. Will do.

File: url/Slug.ts [hunk 3]
```diff
@@ -1,27 +1,16 @@
...
-     * Constructs a new instance
-     *
-     * @param {Array.<Array.<*>>} transforms
-     * @param {?RegExp} sanitize
+     * The sanitizing regexp, that removes all unwanted characters
      */
-    constructor (transforms = [], sanitize = null)
+    private readonly sanitize : RegExp;
+
+    constructor (transforms : SlugCharacterReplacementMap = [], sanitize : null|RegExp= null)
```
Comment (apfelbox) [inline url/Slug.ts]:
```suggestion
    constructor (transforms : SlugCharacterReplacementMap = [], sanitize : null|RegExp = null)
```

File: url/Slug.ts [hunk 4]
```diff
@@ -1,27 +1,16 @@
...
-         */
         this.transforms = [];
-
-        /**
-         * The sanitizing regexp, that removes all unwanted characters
-         *
-         * @private
-         * @type {RegExp}
-         */
         this.sanitize = null !== sanitize
```
Comment (apfelbox) [inline url/Slug.ts]: `private`?

File: url/Slug.ts [hunk 5]
```diff
@@ -1,27 +1,16 @@
...
-    constructor (transforms = [], sanitize = null)
+    private readonly sanitize : RegExp;
+
+    constructor (transforms : SlugCharacterReplacementMap = [], sanitize : null|RegExp= null)
     {
-        /**
-         * @private
-         * @type {Array.<Array.<RegExp, string>>}
-         */
         this.transforms = [];
```
Comment (apfelbox) [inline url/Slug.ts]:
```suggestion
        private this.transforms = [];
```

File: url/Slug.ts (hunk 4 above)
Comment (apfelbox) [reply url/Slug.ts]:
```suggestion
        private this.sanitize = null !== sanitize
```

File: url/Slug.ts [hunk 6]
```diff
@@ -40,13 +29,13 @@ export default class Slug
         for (let i = 0; i < transforms.length; i++)
         {
             let from = transforms[i][0];
-            const flags = typeof from.flags === "string"
-                ? from.flags
-                : "";
 
-            if (-1 === flags.indexOf("g"))
+            if (from instanceof RegExp)
```
Comment (apfelbox) [inline url/Slug.ts]:
```suggestion
            if (from instanceof RegExp && -1 === from.flags.indexOf("g"))
```

Comment (apfelbox) [reply url/Slug.ts]: And remove the if below.

File: url/Slug.ts (hunk 4 above)
Comment (keichinger) [reply url/Slug.ts]: At the declaration level, not at the assignment :)

**Pull request merged**